File: src/mnemonic/converter/tlg/tlg5.py
"""TLG5デコーダーモジュール

TLG5形式の画像をデコードする機能を提供する。
TLG5はLZSS圧縮を使用したカラープレーン分離方式の画像フォーマット。
"""

from dataclasses import dataclass
from typing import Protocol

from PIL import Image

from mnemonic.converter.tlg.lzss import LZSSDecoder


class TLG5DecoderProtocol(Protocol):
    """TLG5デコーダーインターフェース

    TLG5形式の画像をデコードするためのプロトコル定義。
    """

    def decode(self, data: bytes) -> Image.Image:
        """TLG5形式のバイト列をデコードする

        Args:
            data: TLG5形式の画像バイト列

        Returns:
            デコードされたPIL.Imageオブジェクト

        Raises:
            ValueError: 不正なTLG5データの場合
        """
        ...


@dataclass(frozen=True)
class TLG5Header:
    """TLG5ヘッダー情報

    TLG5画像ファイルのヘッダーから読み取った情報を保持する不変データクラス。

    Attributes:
        width: 画像の幅（ピクセル）
        height: 画像の高さ（ピクセル）
        colors: 色数（3=RGB、4=RGBA）
        block_height: ブロックの高さ（ピクセル）
    """

    width: int
    height: int
    colors: int
    block_height: int


class TLG5Decoder:
    """TLG5画像デコーダー

    TLG5形式の画像をデコードしてPIL.Imageオブジェクトに変換する。
    TLG5はLZSS圧縮を使用し、カラープレーン分離方式でデータを格納する。

    TLG5形式の構造:
    - ヘッダー: マジック(11) + 色深度(1) + width(4) + height(4) + block_height(4)
    - データ: 各ブロックごとに mark(1) + block_size(4) + compressed_data
    - 各ブロックは各色チャンネル(B,G,R,A)ごとに格納
    - デルタエンコーディング（行差分）を使用
    """

    MAGIC: bytes = b"TLG5.0\x00raw\x1a"
    """TLG5形式のマジックバイト"""

    HEADER_SIZE: int = 24
    """ヘッダーサイズ: マジック(11) + 色深度(1) + width(4) + height(4) + block_height(4)"""

    def __init__(self) -> None:
        """TLG5デコーダーを初期化する"""
        self._lzss = LZSSDecoder()

    def is_valid(self, data: bytes) -> bool:
        """指定されたデータがTLG5形式かどうかを判定する

        Args:
            data: 判定対象のバイト列

        Returns:
            TLG5形式の場合True、そうでない場合False
        """
        return data.startswith(self.MAGIC)

    def parse_header(self, data: bytes) -> TLG5Header:
        """TLG5ヘッダーを解析する

        Args:
            data: TLG5形式の画像バイト列

        Returns:
            解析されたヘッダー情報

        Raises:
            ValueError: TLG5形式でないデータ、またはデータが短すぎる場合
        """
        if not self.is_valid(data):
            raise ValueError("TLG5形式ではありません")

        if len(data) < self.HEADER_SIZE:
            raise ValueError("データが短すぎます")

        offset = len(self.MAGIC)

        # 色深度: 24=RGB(3色), 32=RGBA(4色)
        color_depth = data[offset]
        colors = 4 if color_depth == 32 else 3
        offset += 1

        # 画像サイズ（リトルエンディアン）
        width = int.from_bytes(data[offset : offset + 4], "little")
        offset += 4
        height = int.from_bytes(data[offset : offset + 4], "little")
        offset += 4

        # ブロック高さ（リトルエンディアン）
        block_height = int.from_bytes(data[offset : offset + 4], "little")

        return TLG5Header(
            width=width,
            height=height,
            colors=colors,
            block_height=block_height,
        )

    def decode(self, data: bytes) -> Image.Image:
        """TLG5形式のバイト列をデコードする

        Args:
            data: TLG5形式の画像バイト列

        Returns:
            デコードされたPIL.Imageオブジェクト

        Raises:
            ValueError: TLG5形式でないデータ、または不正なデータの場合
        """
        if not self.is_valid(data):
            raise ValueError("TLG5形式ではありません")

        header = self.parse_header(data)
        width = header.width
        height = header.height
        colors = header.colors
        block_height = header.block_height

        # ブロック数を計算
        block_count = (height + block_height - 1) // block_height

        # 各チャンネルのピクセルデータを格納
        # TLG5はBGRA順で格納される
        channels: list[bytearray] = [bytearray(width * height) for _ in range(colors)]

        # データ読み取り開始位置
        offset = self.HEADER_SIZE

        # 各ブロックを処理
        for block_idx in range(block_count):
            # このブロックの開始行と行数
            block_y_start = block_idx * block_height
            block_rows = min(block_height, height - block_y_start)
            block_pixel_count = width * block_rows

            # 各チャンネルのブロックデータを読み取り
            for channel_idx in range(colors):
                # ブロックヘッダーを読み取り: mark(1) + block_size(4)
                if offset + 5 > len(data):
                    raise ValueError("ブロックデータが不完全です")

                # mark は通常0（未使用フラグ）
                offset += 1

                block_size = int.from_bytes(data[offset : offset + 4], "little")
                offset += 4

                # 圧縮データを読み取り
                if offset + block_size > len(data):
                    raise ValueError("ブロックデータが不完全です")

                compressed_data = data[offset : offset + block_size]
                offset += block_size

                # LZSS解凍
                decompressed = self._lzss.decode(compressed_data, block_pixel_count)

                # デルタエンコーディングを復元
                self._apply_delta_decoding(
                    channels[channel_idx],
                    decompressed,
                    width,
                    block_y_start,
                    block_rows,
                )

        # BGRAからRGBA/RGBに変換してPIL Imageを作成
        return self._create_image(channels, width, height, colors)

    def _apply_delta_decoding(
        self,
        channel: bytearray,
        delta_data: bytes,
        width: int,
        y_start: int,
        rows: int,
    ) -> None:
        """デルタエンコーディングされたデータを復元する

        TLG5のデルタエンコーディング:
        - 各行の最初のピクセルは前の行の最初のピクセルとの差分
        - 行内の各ピクセルは同じ行の前のピクセルとの差分

        Args:
            channel: 復元先のチャンネルデータ
            delta_data: デルタエンコーディングされたデータ
            width: 画像の幅
            y_start: ブロックの開始行
            rows: ブロックの行数
        """
        delta_pos = 0

        for row in range(rows):
            y = y_start + row
            row_offset = y * width

            for x in range(width):
                delta_value = delta_data[delta_pos]
                delta_pos += 1

                if x == 0:
                    # 各行の最初のピクセル: 前の行の最初のピクセルとの差分
                    if y == 0:
                        # 最初の行は直接値
                        channel[row_offset] = delta_value
                    else:
                        prev_row_offset = (y - 1) * width
                        channel[row_offset] = (channel[prev_row_offset] + delta_value) & 0xFF
                else:
                    # 行内の他のピクセル: 同じ行の前のピクセルとの差分
                    channel[row_offset + x] = (channel[row_offset + x - 1] + delta_value) & 0xFF

    def _create_image(
        self,
        channels: list[bytearray],
        width: int,
        height: int,
        colors: int,
    ) -> Image.Image:
        """チャンネルデータからPIL Imageを作成する

        TLG5はBGRA順で格納されているため、RGBA/RGB順に変換する。

        Args:
            channels: 各チャンネルのピクセルデータ（BGRA順）
            width: 画像の幅
            height: 画像の高さ
            colors: 色数（3=RGB、4=RGBA）

        Returns:
            作成されたPIL Imageオブジェクト
        """
        pixel_count = width * height

        if colors == 4:
            # BGRA -> RGBA
            rgba_data = bytearray(pixel_count * 4)
            for i in range(pixel_count):
                rgba_data[i * 4] = channels[2][i]  # R <- B channel position 2
                rgba_data[i * 4 + 1] = channels[1][i]  # G <- G channel position 1
                rgba_data[i * 4 + 2] = channels[0][i]  # B <- R channel position 0
                rgba_data[i * 4 + 3] = channels[3][i]  # A <- A channel position 3
            return Image.frombytes("RGBA", (width, height), bytes(rgba_data))
        else:
            # BGR -> RGB
            rgb_data = bytearray(pixel_count * 3)
            for i in range(pixel_count):
                rgb_data[i * 3] = channels[2][i]  # R <- B channel position 2
                rgb_data[i * 3 + 1] = channels[1][i]  # G <- G channel position 1
                rgb_data[i * 3 + 2] = channels[0][i]  # B <- R channel position 0
            return Image.frombytes("RGB", (width, height), bytes(rgb_data))

File: internal/converter/tlg/tlg5.go
package tlg

import (
	"bytes"
	"encoding/binary"
	"errors"
	"fmt"
	"image"
)

// TLG5Magic はTLG5形式のマジックバイト。
var TLG5Magic = []byte("TLG5.0\x00raw\x1a")

// TLG5HeaderSize はTLG5ヘッダーサイズ: マジック(11) + 色深度(1) + width(4) +
// height(4) + block_height(4)。
const TLG5HeaderSize = 24

// TLG5デコード時のセンチネルエラー群。
var (
	ErrTLG5InvalidMagic        = errors.New("TLG5形式ではありません")
	ErrTLG5DataTooShort        = errors.New("データが短すぎます")
	ErrTLG5IncompleteBlockData = errors.New("ブロックデータが不完全です")

	// ErrTLG5InvalidBlockHeight はblock_heightが0以下の場合のエラー。
	//
	// why not: Python参照実装はblock_height==0をガードせずブロック数計算
	// (height + block_height - 1) // block_heightでZeroDivisionErrorを
	// 未捕捉のまま送出する（呼び出し元プロセスがクラッシュする）。mnemonicは
	// 信頼できないゲームアセットを処理するCLIであり、不正なTLG5ファイル1件で
	// ビルド全体がpanicで落ちるのは受容できないため、Go版はここでガードし
	// 通常のerrorとして返す。
	ErrTLG5InvalidBlockHeight = errors.New("ブロック高さが不正です")

	// ErrTLG5InvalidDimensions はwidth/heightが0以下、または許容上限
	// （TLG5MaxDimension・TLG5MaxPixelCount）を超える場合のエラー。
	//
	// why not: レビュー指摘の回帰防止。width=height=0xFFFFFFFFのような
	// 24バイトの不正な最小ヘッダーに対し、旧実装はmake([]byte, width*height)
	// をチャンネル数分呼び出す前に何の検証も行っていなかった。
	// ConversionManagerのワーカーgoroutineはrecoverを持たないため、この
	// makeslice panicはビルド処理全体を巻き込んでクラッシュさせる
	// （実際にpipeline経由で再現された）。信頼できないゲームアセットを
	// 処理するCLIとして、確保前にサイズを検証し通常のerrorとして拒否する。
	ErrTLG5InvalidDimensions = errors.New("画像サイズが不正です")

	// ErrTLG5InvalidColorDepth はTLG5ヘッダーの色深度バイトが既知のいずれの
	// 表記（3/24=RGB、4/32=RGBA）にも一致しない場合のエラー。
	ErrTLG5InvalidColorDepth = errors.New("色深度が不正です")

	// ErrTLG5InvalidBlockMark はブロックのmarkバイトが既知の値
	// （0=圧縮、1=格納）以外の場合のエラー。
	ErrTLG5InvalidBlockMark = errors.New("不正なブロックマークです")
)

const (
	// TLG5MaxDimension は幅・高さそれぞれの許容上限。実在するゲームアセットの
	// TLG5画像がこれを超える例は無い。上限が無いとwidth*heightの計算や
	// 後続のバイト列確保がオーバーフロー・OOM相当の巨大確保を引き起こす
	// （例: width=height=0xFFFFFFFFの最小ヘッダーでのmakeslice panic）。
	TLG5MaxDimension = 1 << 16 // 65536

	// TLG5MaxPixelCount はwidth*heightの許容上限。RGBA(4ch)ではcolors数分の
	// バイト列（各width*heightバイト）を確保するため、このキャップは概ね
	// 256MB (67,108,864 pixels * 4ch * 1byte) のメモリ確保上限に相当する。
	// 65536×65536のような上限ぎりぎりの正方形1辺だけでは約17GBの確保に
	// 到達しうるため、辺の上限とは別に面積（pixel数）でも制限する。
	TLG5MaxPixelCount = 64 * 1024 * 1024 // 64Mピクセル

	// TLG5ブロックのmarkバイトの意味（krkrz TVPLoadTLG5ローダー準拠）。
	tlg5BlockMarkCompressed = 0 // LZSS圧縮ブロック
	tlg5BlockMarkStored     = 1 // 非圧縮(格納)ブロック
)

// TLG5Header はTLG5画像ファイルのヘッダーから読み取った情報を保持する不変値。
type TLG5Header struct {
	Width       int
	Height      int
	Colors      int // 3=RGB、4=RGBA
	BlockHeight int
}

// TLG5Decoder はTLG5形式（LZSS圧縮・カラープレーン分離方式）の画像を
// デコードする。
type TLG5Decoder struct {
	lzss *LZSSDecoder
}

// NewTLG5Decoder はTLG5Decoderを初期化する。
func NewTLG5Decoder() *TLG5Decoder {
	return &TLG5Decoder{lzss: NewLZSSDecoder()}
}

// IsValid はdataがTLG5形式のマジックバイトを持つかどうかを判定する。
func (d *TLG5Decoder) IsValid(data []byte) bool {
	return bytes.HasPrefix(data, TLG5Magic)
}

// ParseHeader はTLG5ヘッダーを解析する。
func (d *TLG5Decoder) ParseHeader(data []byte) (TLG5Header, error) {
	if !d.IsValid(data) {
		return TLG5Header{}, ErrTLG5InvalidMagic
	}

	if len(data) < TLG5HeaderSize {
		return TLG5Header{}, ErrTLG5DataTooShort
	}

	offset := len(TLG5Magic)

	colorDepth := data[offset]
	colors, colorErr := channelCountFromColorDepth(colorDepth)
	if colorErr != nil {
		return TLG5Header{}, colorErr
	}
	offset++

	width := int(binary.LittleEndian.Uint32(data[offset : offset+4]))
	offset += 4
	height := int(binary.LittleEndian.Uint32(data[offset : offset+4]))
	offset += 4
	blockHeight := int(binary.LittleEndian.Uint32(data[offset : offset+4]))

	return TLG5Header{Width: width, Height: height, Colors: colors, BlockHeight: blockHeight}, nil
}

// channelCountFromColorDepth はTLG5ヘッダーの色深度バイトをチャンネル数へ
// 変換する。
//
// why not: krkrz（吉里吉里ネイティブC++実装）のTVPLoadTLG5ローダーは、この
// バイトを「チャンネル数」（3=RGB、4=RGBA）として解釈する。一方、Python
// 参照実装（mnemonic feat/exe-icon-extraction）は「24=RGB、32=RGBA」という
// ビット深度風の値としてのみ解釈しており、colors=4のように実機で生成された
// RGBA TLG5ファイルを誤って3チャンネルとしてデコードしチャンネル境界が
// ズレる(desync)欠陥を引き継いでいた。どちらの表記のTLG5ファイルも実在し
// うる（レビューア差分フィクスチャ生成器はPython流の24/32を書き出す）ため、
// 本実装は両方の表記を受け付けるハイブリッド解釈とし、それ以外の値は
// ErrTLG5InvalidColorDepthとして拒否する。
func channelCountFromColorDepth(colorDepth byte) (int, error) {
	switch colorDepth {
	case 3, 24:
		return 3, nil
	case 4, 32:
		return 4, nil
	default:
		return 0, fmt.Errorf("%w: %d", ErrTLG5InvalidColorDepth, colorDepth)
	}
}

// validateDimensions はheaderのWidth/HeightがTLG5MaxDimension・
// TLG5MaxPixelCountの範囲内にあることを検証する。呼び出し元は、この検証を
// width/height由来のバイト列確保（channels/image.Image作成）より前に必ず
// 行うこと。
func validateDimensions(header TLG5Header) error {
	if header.Width <= 0 || header.Height <= 0 {
		return fmt.Errorf("%w: %dx%d", ErrTLG5InvalidDimensions, header.Width, header.Height)
	}

	if header.Width > TLG5MaxDimension || header.Height > TLG5MaxDimension {
		return fmt.Errorf("%w: %dx%d", ErrTLG5InvalidDimensions, header.Width, header.Height)
	}

	// width・heightは共にTLG5MaxDimension(65536)以下であることが上の
	// チェックで保証されるため、この乗算は最大でも65536*65536=4,294,967,296
	// となり64bit intの範囲でオーバーフローしない。
	if header.Width*header.Height > TLG5MaxPixelCount {
		return fmt.Errorf("%w: %dx%d", ErrTLG5InvalidDimensions, header.Width, header.Height)
	}

	return nil
}

// Decode はTLG5形式のバイト列をデコードし、image.Imageを返す。
//
// 戻り値の具象型はcolors(3=RGBはimage.RGBA、4=RGBAはimage.NRGBA)に応じて
// 異なる（詳細はcreateImageFromChannelsのwhy notコメントを参照）。
func (d *TLG5Decoder) Decode(data []byte) (image.Image, error) {
	if !d.IsValid(data) {
		return nil, ErrTLG5InvalidMagic
	}

	header, err := d.ParseHeader(data)
	if err != nil {
		return nil, err
	}

	// why: width/height由来のバイト列確保(下のchannels := make(...))より
	// 前に必ず検証する。この順序を破ると不正な巨大寸法でmakesliceが
	// panicする（レビュー指摘の回帰防止）。
	if dimErr := validateDimensions(header); dimErr != nil {
		return nil, dimErr
	}

	if header.BlockHeight <= 0 {
		return nil, ErrTLG5InvalidBlockHeight
	}

	width, height, colors, blockHeight := header.Width, header.Height, header.Colors, header.BlockHeight
	blockCount := (height + blockHeight - 1) / blockHeight

	// TLG5はBGRA順で格納される
	channels := make([][]byte, colors)
	for i := range channels {
		channels[i] = make([]byte, width*height)
	}

	// ブロックサイズ配列の後からブロックデータが始まる
	offset := TLG5HeaderSize + blockCount*4

	for blockIdx := range blockCount {
		blockYStart := blockIdx * blockHeight
		blockRows := min(blockHeight, height-blockYStart)
		blockPixelCount := width * blockRows

		for channelIdx := range colors {
			// 各チャンネルデータにはmark(1) + size(4)のヘッダーがある
			if offset+5 > len(data) {
				return nil, ErrTLG5IncompleteBlockData
			}

			mark := data[offset]
			offset++

			channelSize := int(binary.LittleEndian.Uint32(data[offset : offset+4]))
			offset += 4

			if offset+channelSize > len(data) {
				return nil, ErrTLG5IncompleteBlockData
			}

			blockData := data[offset : offset+channelSize]
			offset += channelSize

			decompressed, blockErr := decodeBlock(d.lzss, mark, blockData, blockPixelCount)
			if blockErr != nil {
				return nil, blockErr
			}

			applyDeltaDecoding(channels[channelIdx], decompressed, width, blockYStart, blockRows)
		}
	}

	return createImageFromChannels(channels, width, height, colors), nil
}

// decodeBlock はTLG5ブロック内の1チャンネル分のデータをmarkバイトに従って
// 復元する。
//
// markの意味はkrkrz（吉里吉里ネイティブC++実装）のTVPLoadTLG5ローダーに
// 準拠する: 0=LZSS圧縮ブロック、1=非圧縮の格納ブロック（エンコーダはLZSS
// 圧縮が展開後より大きくなる場合に格納ブロックを使う）。
//
// why not(格納ブロックとスライド辞書の関係、未検証の前提): krkrzの実装では
// 格納ブロックのデータもスライド辞書へ書き戻され、後続の圧縮ブロックの
// バックリファレンスが辞書状態を前提にしている可能性がある。しかし本実装
// （移植元のPython参照実装feat/exe-icon-extractionを含む）はLZSSDecoder.
// Decodeをチャンネル×ブロックの圧縮データチャンクごとに独立して呼び出し、
// スライド辞書を毎回ゼロ初期化する設計になっている（辞書状態はチャンク
// 境界をまたいで一切引き継がれない。圧縮ブロック同士の間でも共有しない）。
// この設計のもとでは、格納ブロックについてのみ辞書へ書き戻す処理を追加
// しても後続チャンクの解凍結果には数学的に影響しない（次のLZSSDecoder.
// Decode呼び出しは常にゼロ初期化された辞書から始まるため）。したがって
// 格納ブロックは単純にバイト列をそのままピクセル値として採用するのみで
// よい、という保守的な解釈を採用する。
//
// 実機のkrkrz実装が本当にチャンク単位で辞書をリセットしているのか、
// この一致がTLG5フォーマットの「正しい」仕様なのか、それとも参照実装の
// 簡略化がたまたま無害だっただけなのかは、参照実装の入手範囲では検証
// できていない。この不確実性を明記した上で、現状の挙動をテストで
// ピン留めする。
func decodeBlock(lzss *LZSSDecoder, mark byte, blockData []byte, pixelCount int) ([]byte, error) {
	switch mark {
	case tlg5BlockMarkCompressed:
		decompressed, decErr := lzss.Decode(blockData, pixelCount)
		if decErr != nil {
			return nil, fmt.Errorf("%w: %w", ErrTLG5IncompleteBlockData, decErr)
		}

		return decompressed, nil
	case tlg5BlockMarkStored:
		if len(blockData) != pixelCount {
			return nil, fmt.Errorf("%w: 格納ブロックのサイズ不一致 got=%d want=%d", ErrTLG5IncompleteBlockData, len(blockData), pixelCount)
		}

		return blockData, nil
	default:
		return nil, fmt.Errorf("%w: %d", ErrTLG5InvalidBlockMark, mark)
	}
}

// applyDeltaDecoding はデルタエンコーディングされたデータを復元する。
//
// TLG5のデルタエンコーディング: 各ピクセル = (水平累積 + delta + 上のピクセル)
// & 0xFF。水平累積は各行でリセットされ、最初の行（y=0）では上のピクセル=0。
func applyDeltaDecoding(channel, deltaData []byte, width, yStart, rows int) {
	deltaPos := 0

	for row := range rows {
		y := yStart + row
		rowOffset := y * width
		hasPrevRow := y > 0
		prevRowOffset := (y - 1) * width

		// 各行で水平累積をリセット。byte型の加算はGoの仕様上mod 256で
		// wrapするため、Python版の `& 0xFF` に相当する挙動になる。
		var horizontalAccum byte

		for x := range width {
			horizontalAccum += deltaData[deltaPos]
			deltaPos++

			var upperPixel byte
			if hasPrevRow {
				upperPixel = channel[prevRowOffset+x]
			}

			channel[rowOffset+x] = horizontalAccum + upperPixel
		}
	}
}

// createImageFromChannels はチャンネルデータ（BGRA順）からimage.Imageを
// 作成する。
//
// why not: colors==4(RGBA)ではimage.NRGBAを、colors==3(RGB)ではimage.RGBAを
// 使い分ける。両者ともR/G/B/Pix配置は同一（差はAの意味論のみ）だが、
// image.Image.ColorModel()の型はconverter.imageHasAlpha（呼び出し側
// internal/converter/image.go）がPILのRGBA modeに相当する「全画素不透明でも
// アルファ有り扱い」をNRGBAModelのみで判定する根拠にしている。RGB(3チャンネル)
// 由来の画像はA=255固定で意味論上アルファを持たないため、NRGBAModelにすると
// 常にアルファ有りと誤判定されWebP出力時に不要なロスレス強制が起きる。
// image.RGBAはA=255時は非プリマルチプライ相当と数値上一致するため、この
// 使い分けはピクセル値の意味を変えない。
func createImageFromChannels(channels [][]byte, width, height, colors int) image.Image {
	rect := image.Rect(0, 0, width, height)
	pixelCount := width * height

	if colors == 4 {
		img := image.NewNRGBA(rect)
		for i := range pixelCount {
			px := i * 4
			img.Pix[px] = channels[2][i]   // R <- チャンネル2(格納順ではBGRAのR位置)
			img.Pix[px+1] = channels[1][i] // G
			img.Pix[px+2] = channels[0][i] // B <- チャンネル0
			img.Pix[px+3] = channels[3][i] // A
		}

		return img
	}

	img := image.NewRGBA(rect)
	for i := range pixelCount {
		px := i * 4
		img.Pix[px] = channels[2][i]   // R <- チャンネル2(格納順ではBGRのR位置)
		img.Pix[px+1] = channels[1][i] // G
		img.Pix[px+2] = channels[0][i] // B <- チャンネル0
		img.Pix[px+3] = 255
	}

	return img
}
